test/Microsoft.Health.Fhir.Shared.Tests.E2E/Rest/Audit/AuditTests.cs

```diff
@@ -380,6 +380,139 @@ public async Task GivenAResource_WhenNotAuthorized_ThenAuditLogEntriesShouldBeCr
+                return;
+            }
+
+            List<(string expectedActions, string expectedPathSegments, HttpStatusCode? expectedStatusCodes, ResourceType? resourceType)> expectedList = new List<(string, string, HttpStatusCode?, ResourceType?)>
+            {
+                ("transaction", string.Empty, null, null),
+                ("create", "Patient", null, null),
+                ("create", "Patient", HttpStatusCode.Created, ResourceType.Patient),
+                ("read", "Patient/12345", null, null),
+                ("read", "Patient/12345", HttpStatusCode.NotFound, ResourceType.OperationOutcome),
+                ("transaction", string.Empty, HttpStatusCode.NotFound, ResourceType.OperationOutcome),
+            };
+
+            var requestBundle = Samples.GetJsonSample("Bundle-TransactionForRollBack");
+
+            await ExecuteAndValidateTransaction(
+              async () =>
+              {
+                  FhirResponse<OperationOutcome> result = null;
+
+                  try
+                  {
+                      await _client.PostBundleAsync(requestBundle.ToPoco<Bundle>());
+                  }
+                  catch (FhirException ex)
+                  {
+                      result = ex.Response;
+                  }
+
+                  // The request should have failed.
+                  Assert.NotNull(result);
+
+                  return result;
+              },
+              expectedList);
+        }
+
+        private async Task ExecuteAndValidateTransaction<T>(Func<Task<FhirResponse<T>>> action, List<(string, string, HttpStatusCode?, ResourceType?)> expectedList)
+           where T : Resource
+        {
+            if (!_fixture.IsUsingInProcTestServer)
+            {
+                // This test only works with the in-proc server with customized middleware pipeline.
+                return;
+            }
+
+            FhirResponse<T> response = null;
+
+            response = await action();
+
+            string correlationId = response.Headers.GetValues(RequestIdHeaderName).FirstOrDefault();
+
+            Assert.NotNull(correlationId);
+
+            string expectedAppId = TestApplications.ServiceClient.ClientId;
+
+            IReadOnlyList<AuditEntry> auditList = _auditLogger.GetAuditEntriesByCorrelationId(correlationId);
+
+            Assert.Equal(expectedList.Count, auditList.Count);
+
+            for (int iter = 0; iter < auditList.Count; iter++)
+            {
+                if (iter == 0)
+                {
+                    // Validates the first entry is logging transaction is getting executed
+                    ValidateExecutingAuditEntry(auditList[0], expectedList[0].Item1, new Uri($"http://localhost/{expectedList[0].Item2}"), correlationId, expectedAppId, ExpectedClaimKey);
+                    continue;
+                }
+
+                if (iter == auditList.Count - 1)
+                {
```

Comment:
if block (iter == auditList.Count - 1 || iter % 2 == 0) can be combined together

Reply:
yes they can be combined together. I did this way in order to make the last item stand out so that I can document the goal of first and last element of audit entry. It is done this way for readability purpose. What do you think?